.pre-commit-config.yaml

```diff
@@ -1,33 +1,77 @@
+# See https://pre-commit.com for more information
+# See https://pre-commit.com/hooks.html for more hooks
+
+minimum_pre_commit_version: '2.8.2'
+
+default_language_version:
+  python: python3
+
+default_stages: [commit]
+
+
 repos:
+  # General file checks and fixers
   - repo: https://github.com/pre-commit/pre-commit-hooks
-    rev: v3.4.0
+    rev: v4.1.0
     hooks:
       - id: mixed-line-ending
-        name: Normalize mixed line endings
+        name: "Normalize mixed line endings"
         args: [--fix=lf]
+      - id: fix-byte-order-marker
+        name: "Remove Unicode BOM"
+      - id: file-contents-sorter
+        name: "Sort codespell ignore list"
+        files: '.codespell/ignore-words.txt'
+
+      - id: check-case-conflict
+        name: "Check for case conflicts"
+      - id: check-merge-conflict
+        name: "Check for merge conflict markers"
+      - id: check-executables-have-shebangs
+        name: "Check that executables have shebangs"
+      - id: check-shebang-scripts-are-executable
+        name: "Check that shebangs are executable"
+
+      - id: check-vcs-permalinks
+        name: "Check that VCS links are permalinks"
+
+      - id: check-ast
+        name: "Check Python AST"
+      - id: check-json
+        name: "Check JSON"
+      - id: check-toml
+        name: "Check TOML"
+      - id: check-yaml
+        name: "Check YAML"
```

Comment:
I took pains to be pretty conservative and only enabled a limited number of built-in checks after careful investigation and testing that were not just style-related and had an articulatable benefit for this repo, not enabling any that required fixing a few existing files in the repo, none that required dealing with any existing false positives throughout all past PEPs (except for the new PEP link check you specifically requested), and not adding any code checkers, linters or formatters (beyond simply verifying valid AST).

* The BOM check avoids a spurious Unicode BOM injected into text files, as older versions of Notepad do by default and some other text editors can do if set incorrectly.
* The file contents sorter keeps the codespell ignore list sorted and easy to scan. It only runs if that file is modified.
* The case conflict checks for duplicate files and directories that differ only by case (on Unix and Linux), which will cause problems for Git and users on non-case-sensitive filesystems (Windows and most macOS). Not common, but difficult to fix once it gets committed.
* The shebang checks detect instances of the executable bit and the shebang not matching, which is pretty easy to forget especially for users on non-POSIX systems and is fairly easy to fix but hard for reviewers to spot; this I've noticed on multiple PRs just in the past couple weeks (as you know).
* The VCS permalinks check we've discussed; it detects instances of non-perma links to specific lines of files that are likely to quickly become stale; we could extend it to check all VCS links (but that might cause a few false positives, since occasionally there could be reasons for wanting to link the "current" version rather than a specific one).
* The file validation checkers only run if a file with that format is added or modified and contains a critical error (typo, mistake, etc) that would prevent it from being parsed at all, which is certainly quite possible if users are not being careful. In the unlikely event that a deliberately invalid file might be added (so far, no cases of such), it could simply be added to `exclude` for that check and the check won't run at all on it; so far none of the files added are false positives for this.

The extra checks cost users nothing unless they explicitly choose to install our pre-commit hooks _and_ edit a file the checks validate, and even in that case only add a couple extra seconds total once per commit (and on the CI lint run, which is still the fasted CI of the three by several times over). False positives seem rather unlikely, given none of them I ended up adding produced any on the repo's entire history (I rejected those that did).